File: javaCalc/src/main/java/calculator/Calculator.java
package calculator;

import calculator.operations.*;
import calculator.service.Converter;
import calculator.service.RomanConverter;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;
import java.util.regex.Pattern;

public class Calculator {
    private final Converter converter;
    private final Set<Operation> operations;
    private final int min;
    private final int max;

    public Calculator(Set<Operation> operations) {
        this(null, operations);
    }

    public Calculator(Converter converter, Set<Operation> operations) {
        this(converter, operations, 1, 10);
    }

    public Calculator(Converter converter, Set<Operation> operations, int min, int max) {
        this.converter = converter;
        this.operations = operations;
        this.min = min;
        this.max = max;
    }

    public Converter getConverter() {
        return converter;
    }

    public Set<Operation> getOperations() {
        return operations;
    }

    public int getMin() {
        return min;
    }

    public int getMax() {
        return max;
    }

    public static void main(String[] args) {
        new Calculator(RomanConverter.getInstance(), createBasicOperations()).run();
    }

    private static Set<Operation> createBasicOperations() {
        HashSet<Operation> operations = new HashSet<>();
        operations.add(new Plus());
        operations.add(new Minus());
        operations.add(new Multiply());
        operations.add(new Div());

        return operations;
    }

    private void run() {
        Scanner scanner = new Scanner(System.in);
        while (true) {
            try {
                System.out.println(calc(scanner.nextLine()));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public String calc(String expresion) {
        for (Operation operation : getOperations()) {
            if (expresion.contains(operation.getOperator())) {
                String[] numbers = Arrays.stream(expresion.split(Pattern.quote(operation.getOperator())))
                        .map(String::trim)
                        .toArray(String[]::new);

                operation.checkParametersCount(numbers.length);

                boolean needConversion = isNeedConversion(numbers);

                Integer[] arabicNumbers = needConversion ? convertNumbers(numbers) : parseNumbers(numbers);

                checkIsValidNumbers(arabicNumbers);

                Integer result = operation.apply(arabicNumbers);

                return needConversion ? getConverter().convert(result) : result.toString();
            }
        }

        throw new UnsupportedOperationException("неверное выражение, не найден оператор\n" +
                "доступные операторы: " + getOperations());
    }

    private Integer[] parseNumbers(String[] numbers) {
        return Arrays.stream(numbers)
                .map(Integer::parseInt)
                .toArray(Integer[]::new);
    }

    private Integer[] convertNumbers(String[] numbers) {
        for (String number : numbers) {
            if (!converter.isValid(number)) {
                throw new IllegalArgumentException("все числа должны использовать один формат обозначения(Римские, Арабские и т.д.)");
            }
        }

        return Arrays.stream(numbers)
                .map(getConverter()::convert)
                .toArray(Integer[]::new);
    }

    private void checkIsValidNumbers(Integer[] numbers) {
        for (Integer number : numbers) {
            if (!isValidNumber(number)) {
                throw new IllegalArgumentException(String.format("число должно быть в пределах [%d, %d]", getMin(), getMax()));
            }
        }
    }

    private boolean isNeedConversion(String... numbers) {
        if (null == getConverter()) {
            return false;
        }

        for (String number : numbers) {
            if (!isArabicNumber(number)) {
                return true;
            }
        }

        return false;
    }

    private boolean isArabicNumber(String number) {
        return number.matches("\\d*");
    }

    private boolean isValidNumber(Integer number) {
        return number >= getMin() && number <= getMax();
    }


}

File: kotlinCalc/src/main/kt/calculator/Calculator.kt
package calculator

import calculator.operations.*
import calculator.service.Converter
import calculator.service.RomanConverter
import java.util.*

class Calculator(
    private val converter: Converter?,
    private val operations: Set<Operation>,
    private val min: Int = 1,
    private val max: Int = 10
) {

    constructor(operations: Set<Operation>) : this(null, operations)

    fun run() {
        while (true) {
            try {
                println(calc(readLine().toString()))
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    fun calc(expresion: String): String? {
        for (operation in operations) {
            if (expresion.contains(operation.operator)) {
                val numbers = expresion.split(operation.operator).map { it.trim() }

                operation.checkParametersCount(numbers.size)

                val needConversion = isNeedConversion(numbers)

                val arabicNumbers = if (needConversion) convertNumbers(numbers) else parseNumbers(numbers)

                checkIsValidNumbers(arabicNumbers)

                val result = operation.apply(arabicNumbers)

                return if (needConversion) converter!!.convert(result) else result.toString()
            }
        }
        throw UnsupportedOperationException(
            """
                неверное выражение, не найден оператор
                доступные операторы: $operations
                """.trimIndent()
        )
    }

    private fun parseNumbers(numbers: List<String>): List<Int> = numbers.map { it.toInt() }

    private fun convertNumbers(numbers: List<String>): List<Int> {
        return if (numbers.all { converter!!.isValid(it) }) {
            numbers.map { converter!!.convert(it) }
        } else {
            throw IllegalArgumentException("все числа должны использовать один формат обозначения(Римские, Арабские и т.д.)")
        }
    }


    private fun checkIsValidNumbers(numbers: List<Int>) {
        if (numbers.any { !isValidNumber(it) })
            throw java.lang.IllegalArgumentException("""число должно быть в пределах [$min, $max]""")
    }

    private fun isNeedConversion(numbers: List<String>): Boolean {
        converter ?: return false
        return numbers.any { !isArabicNumber(it) }
    }

    private fun isArabicNumber(number: String): Boolean = number.matches("""\d*""".toRegex())

    private fun isValidNumber(number: Int): Boolean = number in min..max

    fun createBasicOperations(): Set<Operation> {
        val operations = HashSet<Operation>()
        operations.add(Plus())
        operations.add(Minus())
        operations.add(Multiply())
        operations.add(Div())
        return operations
    }

    companion object {
        fun createBasicOperations(): Set<Operation> {
            val operations = HashSet<Operation>()
            operations.add(Plus())
            operations.add(Minus())
            operations.add(Multiply())
            operations.add(Div())
            return operations
        }

        fun createCalc() = Calculator(RomanConverter, createBasicOperations())
    }
}

fun main() {
    Calculator(RomanConverter, Calculator.createBasicOperations()).run()
}